### PR TITLE
Lesson 6 and 8: driver fixes and enchancements

### PR DESCRIPTION
**Pull request by MaxGl**
The patch fixes mistakes in the original code, code-style enhancements, simplification of sysfs attributes handling, and char devices support.

Refer to commit messages to get more details.

### CODE REVIEW

File: mpu6050/mpu6050.c [hunk 1]
```diff
@@ -4,60 +4,106 @@
...
 	int accel_values[3];
 	int gyro_values[3];
 	int temperature;
+	unsigned long timestamp;
 };
 
-static struct mpu6050_data g_mpu6050_data;
+static struct i2c_client *mpu6050_drv_client;
+struct mpu6050_data mpu6050_data_samples[SAMPLES_COUNT];
+int mpu6050_current_sample_idx;
```
Comment (an1kh) [inline mpu6050/mpu6050.c]: Maybe declare these variables static?

**Review by an1kh — APPROVED** [review]
Have no objections:
- commits are well-structured, doesn't mess with each other
- commit messages are providing all the needed description
- code looks good

File: mpu6050/mpu6050.c (hunk 1 above)
Comment (MaxGl) [reply mpu6050/mpu6050.c]: Sure, this has to be static. Nicely spotted

Comment (MaxGl) [reply mpu6050/mpu6050.c]: @an1kh fixed

File: mpu6050/mpu6050.c [hunk 2]
```diff
@@ -128,78 +128,72 @@ static struct i2c_driver mpu6050_i2c_driver = {
...
-			   struct class_attribute *attr, char *buf)
-{
-	mpu6050_read_data();
-
-	sprintf(buf, "%d\n", mpu6050_data.gyro_values[Y]);
-	return strlen(buf);
-}
+#define MPU6050_ATTR(_name, _mode, _var) \
+  struct class_ext_attribute mpu6050_class_attr_##_name = \
+  { __ATTR(_name, _mode, mpu6050_show, NULL), _var }
```
Comment (AleksandrBulyshchenko) [inline mpu6050/mpu6050.c]: It worth to write comment with explained intention and dependencies for such macro, which extends kernel functionality and rely on implementation of particular kernel version.

File: mpu6050/mpu6050.c [hunk 3]
```diff
@@ -128,78 +128,72 @@ static struct i2c_driver mpu6050_i2c_driver = {
...
+	NULL
+};
 
-static ssize_t gyro_z_show(struct class *class,
-			   struct class_attribute *attr, char *buf)
+/*
+ * class_* interface as of v4.9 does not provide
+ * a function to register/remove a group of attributes
+ * so implement smth like it on our own
+ */
```
Comment (AleksandrBulyshchenko) [inline mpu6050/mpu6050.c:229]: :+1:

File: mpu6050/mpu6050.c [hunk 4]
```diff
@@ -128,78 +128,72 @@ static struct i2c_driver mpu6050_i2c_driver = {
...
+	void *var;
+};
 
-static ssize_t accel_z_show(struct class *class,
-			    struct class_attribute *attr, char *buf)
+static ssize_t mpu6050_show(struct class *class, struct class_attribute *attr, char *buf)
 {
-	mpu6050_read_data();
+	struct class_ext_attribute *a = container_of(attr, struct class_ext_attribute, attr);
+	int *v = a->var;
```
Comment (AleksandrBulyshchenko) [inline mpu6050/mpu6050.c]: This is trivial case, but anyway variable names could be more verbose.

File: mpu6050/mpu6050.c [hunk 5]
```diff
@@ -239,13 +400,15 @@ static int mpu6050_init(void)
 
 static void mpu6050_exit(void)
 {
+	misc_deregister(&mpu6050_all_dev);
+	misc_deregister(&mpu6050_current_dev);
+	i2c_del_driver(&mpu6050_i2c_driver);
+	pr_info("mpu6050: i2c driver deleted\n");
```
Comment (AleksandrBulyshchenko) [inline mpu6050/mpu6050.c]: Why `i2c_del_driver()` is moved before destroying the sysfs class?
`mpu6050_show() -> mpu6050_read_data()` still can be called.
Will `mpu6050_drv_client` sill valid?

File: mpu6050/mpu6050.c [hunk 6]
```diff
@@ -193,7 +250,91 @@ static void class_remove_group(struct class *class, struct class_attribute **att
...
+
+	mpu6050_seq_print_one(seq, data);
+	return 0;
+}
+
+static int mpu6050_all_seq_show(struct seq_file *seq, void *d)
+{
+	int i;
+
+	for (i = 0; i < SAMPLES_COUNT; i++) {
```
Comment (AleksandrBulyshchenko) [inline mpu6050/mpu6050.c]: If you have already got more than `SAMPLES_COUNT` readings,
you'll output values in wrong order.
In such case you should start from `mpu6050_current_sample_idx + 1` and make a round.

File: mpu6050/mpu6050.c [hunk 7]
```diff
@@ -193,7 +250,91 @@ static void class_remove_group(struct class *class, struct class_attribute **att
...
+
+static int mpu6050_current_open(struct inode *inode, struct file *file)
+{
+	mpu6050_common_open();
+	return single_open(file, mpu6050_current_seq_show, (void*)mpu6050_current_sample_idx);
+}
+
+static int mpu6050_all_open(struct inode *inode, struct file *file)
+{
+	mpu6050_common_open();
```
Comment (AleksandrBulyshchenko) [inline mpu6050/mpu6050.c]: Why do you need to read a new sample when opening `mpu6050_all`?
IMO, only reading `mpu6050_current` should case it.

File: mpu6050/mpu6050.c [hunk 8]
```diff
@@ -14,56 +16,94 @@ enum {
...
 	int gyro_values[3];
 	int temperature;
+	unsigned long timestamp;
 };
 
-static struct mpu6050_data mpu6050_data;
+static struct i2c_client *mpu6050_drv_client;
+struct mpu6050_data mpu6050_data_samples[SAMPLES_COUNT];
+static int mpu6050_current_sample_idx;
+static unsigned long mpu6050_last_access_time;
```
Comment (AleksandrBulyshchenko) [inline mpu6050/mpu6050.c]: IMO, it's better to collect all private global variables above into single static structure

**Review by AleksandrBulyshchenko — APPROVED** [review]
- Good example of commit messages.
- :+1: for `seq_file`.
- Some nitpicks from my side are in inline comments.

**Pull request closed without merging**